### PR TITLE
fix BaseDataset cache='ram' leak with num_workers > 0

### PR DESCRIPTION
**Pull request by raimbekovm**
## Summary

`BaseDataset.cache_images` (detection/segment/pose/obb) populated `self.ims[i]` as a Python list of decoded numpy arrays. With `num_workers > 0`, DataLoader workers either fork (Linux: copy-on-write, eventually duplicated on first touch) or spawn (macOS/Windows: full pickle of every numpy array per worker), so the cache is duplicated per worker and RAM grows unbounded across epochs (same root cause as #9824).

## Fix

Replace the per-image list with a single `torch.uint8` buffer pinned via `share_memory_()` before workers fork:

1. **Two-pass cache build** — pass 1 reads PIL headers (no decode) to compute byte offsets, pass 2 decodes and streams each image directly into the pre-allocated buffer. Avoids holding ~2x peak in a numpy list while building.
2. **Shared memory** — `cache.share_memory_()` makes workers map the same POSIX SHM region instead of duplicating the cache per worker.
3. **Zero-copy `load_image` fast path** — when `img_cache is not None`, return a numpy view into the tensor (no decode, no copy).
4. **Graceful fallback** — wrapped in `try/except (MemoryError, OSError, RuntimeError)` for `/dev/shm` quota in Docker (`--shm-size`).
5. **`safety_margin=1.0`** for `check_cache_ram` — budgets ~2x cache size to absorb streaming/heap-fragmentation overhead during build.

## Validation

2000 synthetic detection images (320–640 px), `YOLODataset(cache='ram', augment=True)`, `DataLoader(num_workers=4)`, 5 epochs, USS measured per epoch:

| Scenario | init USS | e1 | e5 | drift e1→e5 |
|---|---|---|---|---|
| `cache=False, w=0` | 182 MB | 237 MB | 237 MB | +0.1 MB |
| `cache=ram,   w=0` | 826 MB | 826 MB | 826 MB | +0.0 MB |
| `cache=False, w=4` | 829 MB | 321 MB | 142 MB | -180 MB (allocator settles) |
| **`cache=ram,   w=4`** (before) | 602 MB | 530 MB | 1045 MB | **+515 MB** ← leak |
| **`cache=ram,   w=4`** (after) | 660 MB | 324 MB | 396 MB | **+73 MB** ✅ |

Byte-identity: 50/50 `load_image` arrays match between `cache=False` and `cache=ram` (sha256 confirmed).

## Test plan

- [x] `pytest tests/test_python.py::test_grayscale` (4 tasks: detect, obb, segment, pose) — pass
- [x] `pytest tests/test_python.py::test_train_pretrained` — pass
- [x] `pytest tests/test_python.py::test_train_scratch` — pass
- [x] Empirical leak repro: drift drops from +515 MB to +73 MB
- [x] Byte-identity: 50/50 arrays match `cache=False`
- [ ] CI green

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
🚀 This PR overhauls export testing and documentation by adding isolated CI environments for export formats, updating Python/CI support to 3.13, and expanding deployment docs with major new guidance for Hailo, ONNX INT8, RKNN, DEEPX, and Ultralytics Platform integrations.

### 📊 Key Changes
- 🧪 **New isolated export test system in CI**
  - Added a new script, `.github/scripts/create-export-env.py`, to build separate virtual environments for export formats that need special dependencies.
  - Introduced a dedicated **`IsolatedExports`** GitHub Actions job to smoke-test and run export-specific tests outside the shared CI environment.
  - Added a new `--export-env` pytest option so export tests can be filtered by environment.

- ⚙️ **CI modernization and stability improvements**
  - Main CI jobs now use **Python 3.13** by default.
  - Test installation was split into lighter dependency groups like `export-base`, `export-tensorflow`, `export-coreml`, `export-executorch`, and `export-deepx`.
  - Improved handling for older PyTorch versions, ARM runners, caching, retries, and asset preloading.
  - Disabled automatic dependency installs during tests via `YOLO_AUTOINSTALL=false` for more predictable CI behavior.

- 📦 **Packaging and dependency refactor**
  - `pyproject.toml` now supports **Python 3.13** officially.
  - Export dependencies were reorganized into smaller optional groups, making installs more modular and reducing CI conflicts.
  - Added and adjusted dependencies for solutions, ONNX Runtime, OpenVINO INT8, ExecuTorch, DEEPX, TensorFlow, and more.

- 🧠 **Expanded export test coverage**
  - Added tests for:
    - **ONNX INT8 export**
    - **RKNN export**
    - **CoreML RT-DETR export**
    - Export environment registration and smoke validation
  - Improved test isolation by copying models to per-test temp paths to avoid parallel test collisions.

- 🐳 **Docker cleanup improvements**
  - Docker workflows now prune build cache after validation runs.
  - Export Dockerfiles now clean `uv` cache and temp files to reduce image bloat.

- 📚 **Large documentation refresh**
  - Added a full new **Hailo integration guide** for compiling YOLO detection models from ONNX to HEF using Hailo’s external toolchain.
  - Added new **Ultralytics Platform Integrations** docs for importing from **Ultralytics HUB** and **Roboflow**.
  - Added ONNX export reference docs and expanded export argument tables.
  - Updated docs across many integrations:
    - **ONNX**: now documents **INT8 export**
    - **RKNN**: now documents floating-point and INT8 export paths
    - **DEEPX**: clearer branding, platform/runtime support, and updated examples
    - **ExecuTorch**: clarified requirements and output filenames
    - **Comet**: much clearer environment variable configuration docs
    - **TensorRT, QNN, OpenVINO, TF.js, Edge TPU, TorchScript, IMX500, Axelera**: accuracy, compatibility, and wording improvements
  - Replaced placeholder doc metadata with real default descriptions and keywords.

- 🛠️ **Miscellaneous quality improvements**
  - Updated cache actions from `v4` to `v5`.
  - Improved test asset caching, including optional slow assets and CLIP model prefetching.
  - Added new API reference entries and updated author mappings.
  - Cleaned up many small docs issues, examples, links, capitalization, and wording.

### 🎯 Purpose & Impact
- ✅ **More reliable export validation**
  - Export formats with tricky or conflicting dependencies can now be tested in their own environments, reducing flaky CI failures and hidden breakages.

- 🚀 **Better support for newer Python versions**
  - Moving CI to **Python 3.13** helps ensure Ultralytics stays current and compatible with the latest Python ecosystem.

- 🧩 **Cleaner dependency management**
  - Splitting export extras into smaller groups makes development, testing, and installation easier to manage while lowering dependency conflicts.

- 📈 **Improved confidence in export features**
  - New tests for ONNX INT8, RKNN, CoreML, and export environment coverage should catch regressions earlier for users deploying models to edge and production targets.

- 🌍 **Clearer deployment guidance for users**
  - The new and updated docs make it easier for users to understand how to deploy models to specialized hardware like **Hailo**, **Rockchip**, **DEEPX**, and other export targets.

- 🔄 **Smoother platform onboarding**
  - New **Ultralytics Platform** integration docs help users bring data in from HUB or Roboflow more easily, reducing migration friction.

- 🧹 **Leanier CI and Docker workflows**
  - Cache cleanup and improved asset handling should reduce wasted disk space, speed up reruns, and make builds more stable overall.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help streamline review, and an engineer will assist you shortly. Please review the checklist below to support a smooth integration of your changes. 🛠️✨

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/data/base.py [hunk 1]
```diff
@@ -247,34 +261,102 @@ def load_image(self, i: int, rect_mode: bool = True) -> tuple[np.ndarray, tuple[
...
+            if r != 1:
+                w = min(math.ceil(w0 * r), self.imgsz)
+                h = min(math.ceil(h0 * r), self.imgsz)
+            else:
+                w, h = w0, h0
+            return i, (h0, w0), (h, w, self.channels)
+
+        def load(i: int):
+            im = imread(self.im_files[i], flags=self.cv2_flag)
+            h0, w0 = im.shape[:2]
```
Comment (UltralyticsAssistant) [inline ultralytics/data/base.py]: 💡 **MEDIUM**: `imread()` can still return `None` for unreadable or truncated files. In the old path, `load_image()` turned that into a clear `FileNotFoundError`; here we immediately dereference `.shape`, which will surface as an unhandled `AttributeError` during cache build. Please keep the same guard in this helper so RAM caching fails predictably on bad inputs.

File: ultralytics/data/base.py [hunk 2]
```diff
@@ -247,34 +261,102 @@ def load_image(self, i: int, rect_mode: bool = True) -> tuple[np.ndarray, tuple[
...
+                f"{self.prefix}cache='ram' disabled: {type(e).__name__}: {e}. "
+                "Common cause: /dev/shm quota in Docker (raise with --shm-size). Falling back to disk reads."
+            )
+            self.cache = None
+            return
+
+        self.img_cache = cache
+        self.img_offsets = offsets
+        self.img_shapes = shapes
+        self.im_hw0 = hw0
```
Comment (UltralyticsAssistant) [inline ultralytics/data/base.py]: 💡 **MEDIUM**: The shared-RAM branch populates `img_shapes` and `im_hw0`, but leaves `self.im_hw` as the original all-`None` list. Previously `cache='ram'` filled `self.im_hw` during `cache_images()`, so any later code or subclass that reads that documented attribute directly will now regress. Please mirror the old invariant by storing the resized `(h, w)` values there as well.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Nice approach overall: the shared `torch.uint8` buffer and two-pass build should address the per-worker RAM duplication cleanly. I found a few issues to fix before merge: new type syntax that can break older Python interpreters, a missing unreadable-image guard in the new loader path, and `self.im_hw` no longer being initialized in the RAM-cache branch.

💬 Posted 3 inline comments

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The shared-memory approach is a strong fix for the per-worker RAM duplication, but there are two correctness issues to address before merge: the new fast path ignores `rect_mode`, and it returns a mutable view into shared storage that augmentations can modify across workers.

💬 Posted 2 inline comments

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The shared-memory approach is a solid fix for the worker duplication problem, but I found two issues worth addressing before merge: `cache='ram'` appears to regress for `rect_mode=False` datasets, and the broad `OSError` fallback can hide corrupt image errors behind a misleading cache warning.

💬 Posted 2 inline comments

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The shared-memory approach looks like a good fix for the worker duplication issue, but there are two correctness regressions to address before merging: shape metadata is probed with a different decoder than the one used to fill the cache, and the new buffer unconditionally downcasts cached images to `uint8`.

💬 Posted 2 inline comments

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 5

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change looks solid overall. The shared-memory RAM cache addresses the worker duplication issue cleanly, and I didn't find any clear correctness, performance, or maintainability problems in the diff that warrant an inline comment.

**Comment by glenn-jocher** [comment]
Thanks for the clear write-up; the shared-memory approach makes sense, and the main thing still needed here is a focused regression test for `cache='ram'` with `num_workers > 0` plus the constrained shared-memory fallback path so we can review this safely.

**Comment by raimbekovm** [comment]
Superseded by #24673, recreated on a clean branch from main to drop unrelated commits that accumulated on this branch (multiple main merges and auto-format commits inflated the diff to 82 files). The code change is the same and includes the shared-memory allocation improvement.
